Add support for building k0s without embedded binaries

pavansaikrishna78 opened the pull request:
Fixes #6514

This adds the ability to build k0s without embedded binaries through the GitHub Actions workflow.

Changes:
- Add embedded-bins-buildmode input parameter to build-k0s.yml (defaults to 'docker')
- Conditionally skip Docker setup when buildmode is 'none'
- Conditionally skip 'make bindata' when buildmode is 'none'
- Conditionally skip airgap image list generation when buildmode is 'none'
- Add artifact name suffix (-no-embedded-bins) to distinguish builds
- Create build-no-embedded-bins.yml workflow

Local testing shows:
- Build without embedded bins: 111M
- Build with embedded bins: 254M
- Size reduction: 142MB (56.5%)

## Description

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Fixes # (issue)

## Type of change

<!-- check the related options -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [ ] Manual test
- [ ] Auto test added

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

## Checklist

- [ ] My code follows the style [guidelines](https://docs.k0sproject.io/head/contributors/) of this project
- [ ] My commit messages are [signed-off](https://docs.k0sproject.io/head/contributors/github_workflow/)
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

twz123 commented on .github/workflows/build-k0s.yml:
```diff
@@ -81,21 +95,26 @@ jobs:
+          EMBEDDED_BINS_BUILDMODE: "${{ inputs.embedded-bins-buildmode }}"
         run: |
-          make .k0sbuild.docker-image.k0s
+          if [ "$EMBEDDED_BINS_BUILDMODE" != "none" ]; then
```
I'm pretty sure this will break the build, as this will skip building the Docker build image, and then `make build` will fail. I doubt that we need to change anything in the run section of this step, because by setting EMBEDDED_BINS_BUILDMODE=none, the Makefile will already know that it doesn't need to build the embedded binaries and just won't declare them as requisites.

twz123 reviewed (commented):
For reference, this was my experiment: https://github.com/k0sproject/k0s/commit/414a59b5d7f3e3bf7d9accaddb004246b5905931

pavansaikrishna78 committed:
Add support for building k0s without embedded binaries

Fixes #6514

This adds the ability to build k0s without embedded binaries through
the GitHub Actions workflow by adding an embedded-bins-buildmode input
parameter.

Changes:
- Add embedded-bins-buildmode input parameter to build-k0s.yml (defaults to 'docker')
- Pass EMBEDDED_BINS_BUILDMODE to make build, which handles the logic internally
- Conditionally cache embedded binaries only when buildmode is 'docker'
- Conditionally upload airgap image list only when embedded bins are built
- Add artifact name suffix (-no-embedded-bins) to distinguish builds
- Create build-no-embedded-bins.yml workflow for testing builds without embedded bins

The Makefile already handles EMBEDDED_BINS_BUILDMODE=none by:
- Adding the 'noembedbins' build tag
- Not appending bindata to the binary
- Skipping embedded binary builds as prerequisites

Local testing shows:
- Build without embedded bins: 111M
- Build with embedded bins: 254M
- Size reduction: 142MB (56.5%)

Signed-off-by: pavansaikrishna78 <pavansaikrishna78@gmail.com>

pavansaikrishna78 committed:
ci: make SBOM submission continue on error

Add continue-on-error flag to SBOM submission step to prevent
workflow failures when GitHub Dependency Graph is disabled.

This allows the build workflow to complete successfully while
still attempting to submit the SBOM when the feature is available.

Signed-off-by: pavansaikrishna78 <pavansaikrishna78@gmail.com>
